### PR TITLE
Rework Platform-related machinery

### PR DESCRIPTION
**Pull request by FabioLuporini**
This PR adds the Platform class hierarchy (CPU64, Intel64, KNL, ... ARM, ..., Power, ...) so that the compiler can make platform-specific choices when it comes to optimization. 

The nice thing now is that if you want to cross-compile (note: I'm talking about the code that devito generates, not the actual machine code) for a certain target architecture, now you can do it with ``DEVITO_PLATFORM=...`` and for example ``knl`` rather than ``hsw``. On KNL, for example, nested parallel loops are collapsed and nested parallelism is enabled.
By default, we stick to the ``cpu64`` mode, which is essentially what we have in master

This PR also drops now uselss environment variables: DEVITO_DLE_OPTIONS and DEVITO_ISA. In particular, the highest ISA now is bound directly to a Platform, so no need to set it up explicitly.

### CODE REVIEW

File: benchmarks/user/benchmark.py [hunk 1]
```diff
@@ -56,12 +56,11 @@ def option_performance(f):
 
     _preset = {
         # Fixed
-        'O1': {'dse': 'basic', 'dle': 'basic'},
+        'O1': {'dse': 'basic', 'dle': 'advanced'},
```
Comment (mloubout) [inline benchmarks/user/benchmark.py]: really?

File: devito/archinfo.py [hunk 2]
```diff
@@ -25,37 +22,72 @@ def get_cpu_info():
...
-    isa = 'cpp'
-    for i in reversed(known_isas):
-        if any(j.startswith(i) for j in cpu_info['flags']):
-            # Using `startswith`, rather than `==`, as a flag such as 'avx512'
-            # appears as 'avx512f, avx512cd, ...'
-            isa = i
-            break
-    return isa
+def lscpu():
+    p1 = Popen(['lscpu'], stdout=PIPE, stderr=PIPE)
```
Comment (mloubout) [inline devito/archinfo.py]: `lscpu` doesn't exist on osx

File: benchmarks/user/benchmark.py (hunk 1 above)
Comment (FabioLuporini) [reply benchmarks/user/benchmark.py]: yes, I dropped dle basic, it was useless

**Comment by FabioLuporini** [comment]
> lscpu only works on linux system

indeed

> so this gonna break on osx at least

no, or, I mean, this is not the point. That piece of code **attempts** to leverage ``lscpu`` -- we just hope it's available on the underlying system. But if it's not, we proceed silently (see Popen's ``stderr`` gets captured). IOW, the platform autodetection attempts a number of approaches and then if they all fail we still to the fallback ``CPU64``, representing "a generic 64-bit cpu architecture"

**Comment by FabioLuporini** [comment]
all sorted out and green light given offline, so merging

**Comment by FabioLuporini** [comment]
merged.

**Pull request merged**